content/token22-on-chain.md

```diff
@@ -0,0 +1,1483 @@
+
+However, you cannot use any of these types from the `token_interface` module with the regular Anchor `Program` and `Account` wrappers. These new types are used with either the `Interface` or `InterfaceAccount` wrappers. For example, the following would not be valid, and any transactions sent to an instruction using this account deserialization would return an error.
+
+```rust
+// This is invalid, using as an example.
+// Cannot wrap Account over a token_interface::* type.
+pub token_account: Account<'info, token_interface::TokenAccount>
+```
+
+
+# Lab
+
+Now let's get some hands-on experience with the `Token Extension Program` on-chain by implementing a generalized token staking program that will accept both `Token Program` and `Token Extension Program` accounts. As far as staking programs go, this will be a simple implementation with the following design:
+
+* We'll create a stake pool account to hold all the staked tokens. There will only be one staking pool for a given token. The program will own the account. 
+* Every stake pool will have a state account that will hold information regarding the amount of tokens staked in the pool, etc.
+* Users can stake as many tokens as they like, transferring them from their token account to the stake pool.
+* Each user will have a state account created for each pool they stake in. This state account will keep track of how many tokens they have staked in this pool, when they last staked, etc.
+* Users will be minted staking reward tokens upon unstaking. There is no separate claim process required. 
+* We'll determine a user's staking rewards using a simple algorithm.
+* The program will accept both `Token Program` and `Token Extension Program` accounts.
+
+The program will have four instructions: `init_pool`, `init_stake_entry`, `stake`, `unstake`.
+
+This lab will utilize a lot of Anchor and Solana APIs that have been covered previously in this course. We will not spend time explaining some of the concepts we expect you to know. With that said, let's get started.
+
+### 1. Verify Solana/Anchor/Rust Versions
+
+We will be interacting with the `Token Extension` program in this lab and that requires you have solana cli version ≥ `1.18.0`. 
+
+To check your version run:
+```bash
+solana --version
+```
+
+If the version printed out after running `solana --version` is less than `1.18.0` then you can update the [cli version manually](https://docs.solanalabs.com/cli/install). Note, at the time of writing this, you cannot simply run the `solana-install update` command. This command will not update the CLI to the correct version for us, so we have to explicitly download version `1.18.0`. You can do so with the following command:
```

Comment:
It's worth mentioning why (1.18 isn't considered stable yet).

Note the Solana download site was defaulting to `beta` recently, but we will likely fix this soon as `beta` broke a bunch of things.